lib/introspect.js

```diff
@@ -20,32 +20,59 @@ module.exports = function(obj, callback) {
               currentIface = proxy[ifaceName] = {};
 
-              currentIface.addListener = currentIface.on = (function(ifName) {
-                  return function(signame, callback) {
-                      // TODO: check if AddMatch had already been called
-                      // http://dbus.freedesktop.org/doc/api/html/group__DBusBus.html#ga4eb6401ba014da3dbe3dc4e2a8e5b3ef
-                      // An example is "type='signal',sender='org.freedesktop.DBus', interface='org.freedesktop.DBus',member='Foo', path='/bar/foo',destination=':452345.34'" ...
-                      //var match = "type='signal',sender='"+obj.service.name+ ...
-                      
-                      //TODO: add path and interface to path
-                      var match = "type='signal',member='" + signame + "'";
-                      bus.addMatch(match, function(err, result) {
-                          if (err) throw new Error(err);
-                          // TODO: share signal mangling code
-                          var signalFullName = bus.mangle(obj.name, ifName, signame); 
-                          //console.log('trying to listen' +  signalFullName, obj);
-                          bus.signals.on(signalFullName, function(messageBody)
-                          {
+              (function() {
+                  var callbacks = new Array();
+                  var sigHandlers = new Array();
+                  function getSigHandler(callback) {
+                      var index;
+                      if ((index = callbacks.indexOf(callback)) == -1) {
+                          index = callbacks.push(callback) - 1;
+                          sigHandlers[index] = function(messageBody) {
                                callback.apply(null, messageBody)
-                          });
-                      });
+                          }
+                      }
+                      return sigHandlers[index];
                   }
-              })(ifaceName);
 
-              currentIface.removeListener = function() {
-                console.warn("removeListener not implemented for dbus signals");
-              };
-            
+                  currentIface.addListener = currentIface.on = (function(ifName) {
+                      return function(signame, callback) {
+                          // TODO: check if AddMatch had already been called
+                          // http://dbus.freedesktop.org/doc/api/html/group__DBusBus.html#ga4eb6401ba014da3dbe3dc4e2a8e5b3ef
+                          // An example is "type='signal',sender='org.freedesktop.DBus', interface='org.freedesktop.DBus',member='Foo', path='/bar/foo',destination=':452345.34'" ...
+                          //var match = "type='signal',sender='"+obj.service.name+ ...
+
+                          //TODO: add path and interface to path
+                          var match = "type='signal',path='" + obj.name + "',interface='" + ifName + "',member='" + signame + "'";
```

Comment:
maybe share match creation and avoid code duplication? For example, make local function `getMatchRule(signalName)` and use here and from removeListener ( otherwise it's easy to change code here and forget there )